src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectLockFileWatcher.cs

```diff
@@ -0,0 +1,165 @@
+            {
+                return;
+            }
+
+            // If there' no project.json in the project, there's nothing to watch.
+            IProjectTree projectJsonNode = FindProjectJsonNode(newTree);
+            if (projectJsonNode == null)
+            {
+                // project.json may have been deleted.
+                UnregisterFileWatcherIfAny();
+                return;
+            }
+
+            var projectDirectory = Path.GetDirectoryName(_projectServices.Project.FullPath);
+            var projectLockJsonFilePath = Path.ChangeExtension(PathHelper.Combine(projectDirectory, projectJsonNode.Caption), ".lock.json");
+
+            // project.json may have been renamed to {projectName}.project.json. In that case change the file watcher.
+            if (!PathHelper.IsSamePath(projectLockJsonFilePath, _fileBeingWatched))
+            {
+                UnregisterFileWatcherIfAny();
+                RegisterFileWatcher(projectLockJsonFilePath);
+                _fileBeingWatched = projectLockJsonFilePath;
+            }
+        }
+
+        private IProjectTree FindProjectJsonNode(IProjectTree newTree)
+        {
+            IProjectTree projectJsonNode;
+            if (newTree.TryFindImmediateChild("project.json", out projectJsonNode))
+            {
+                return projectJsonNode;
+            }
+
+            var projectName = Path.GetFileNameWithoutExtension(_projectServices.Project.FullPath);
+            if (newTree.TryFindImmediateChild($"{projectName}.project.json", out projectJsonNode))
+            {
+                return projectJsonNode;
+            }
+
+            return null;
+        }
+
+        private void RegisterFileWatcher(string projectLockJsonFilePath)
+        {
+            IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+            if (fileChangeService != null)
+            {
+                int hr = fileChangeService.AdviseFileChange(projectLockJsonFilePath, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Add | _VSFILECHANGEFLAGS.VSFILECHG_Del), this, out _filechangeCookie);
+                ErrorHandler.ThrowOnFailure(hr);
+            }
+        }
+
+        private void UnregisterFileWatcherIfAny()
+        {
+            if (_filechangeCookie != VSConstants.VSCOOKIE_NIL)
+            {
+                IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+                if (fileChangeService != null)
+                {
+                    // There's nothing for us to do if this fails. So ignore the return value.
+                    fileChangeService.UnadviseFileChange(_filechangeCookie);
+                }
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _treeWatcher.Dispose();
```

Comment:
Is there a guarantee that after Dispose no more events are fired? WHat if an event if fired while you are in dispose? You're basically going to re-advise and leak the watch. Imagine a command-line restore while the project is unloading.